# - TAXI NYC - MACHINE LEARNING - DATABRICKS

- Criando Tabela Silver a partir de Amostra de Dados Brutos:


In [0]:
df_bronze = spark.read.table("workspace.default.taxi_nyc_bronze")
df_bronze.write.mode("overwrite").saveAsTable("workspace.default.taxi_nyc_silver")

- Carregando a Tabela Silver e Iniciando a Limpeza:


In [0]:
df_silver = spark.read.table("workspace.default.taxi_nyc_silver")

- Corrigindo Tipos de Dados:


In [0]:
from pyspark.sql.functions import col, to_timestamp

df_silver_tipos_corrigidos = df_silver.withColumn("tpep_pickup_datetime", to_timestamp(col("tpep_pickup_datetime"))) \
                                      .withColumn("tpep_dropoff_datetime", to_timestamp(col("tpep_dropoff_datetime")))

df_silver_tipos_corrigidos.printSchema()

root
 |-- VendorID: long (nullable = true)
 |-- tpep_pickup_datetime: timestamp (nullable = true)
 |-- tpep_dropoff_datetime: timestamp (nullable = true)
 |-- pickup_date: date (nullable = true)
 |-- pickup_day_of_week: long (nullable = true)
 |-- pickup_hour: long (nullable = true)
 |-- trip_duration_minutes: long (nullable = true)
 |-- passenger_count: double (nullable = true)
 |-- trip_distance: double (nullable = true)
 |-- total_amount: double (nullable = true)



- Removendo Dados Nulos:

In [0]:
from pyspark.sql.functions import col

df_silver_filtrado = df_silver_tipos_corrigidos.filter(
    (col("passenger_count") > 0) &
    (col("trip_distance") > 0) &
    (col("total_amount") > 0)
)

- Adicionando Features (pickup_day_of_week,pickup_hour,trip_duration_minutes):

In [0]:
from pyspark.sql.functions import col, dayofweek, hour, round, unix_timestamp

df_silver_enriquecido = df_silver_filtrado.withColumn("pickup_day_of_week", dayofweek(col("tpep_pickup_datetime"))) \
                                          .withColumn("pickup_hour", hour(col("tpep_pickup_datetime")))

duracao_em_segundos = unix_timestamp(col("tpep_dropoff_datetime")) - unix_timestamp(col("tpep_pickup_datetime"))
df_silver_enriquecido = df_silver_enriquecido.withColumn("trip_duration_minutes", round(duracao_em_segundos / 60, 2))

- Salvando os Dados Limpos:

In [0]:
df_silver_enriquecido.write \
  .option("overwriteSchema", "true") \
  .mode("overwrite") \
  .saveAsTable("workspace.default.taxi_nyc_silver")

print("Tabela 'taxi_nyc_silver' atualizada com os dados limpos e enriquecidos.")

Tabela 'taxi_nyc_silver' atualizada com os dados limpos e enriquecidos.


" .option("overwriteSchema", "true") " Força o Delta Lake a ignorar o schema antigo da tabela e usar o novo schema.

- Criando a Camada Gold e Principais Métricas:

In [0]:

df_silver = spark.read.table("workspace.taxi_nyc_schema.taxi_nyc_silver")
from pyspark.sql.functions import count, sum, avg, round, col
# Agrupando os dados e calcular todas asmétricas
df_gold_completa = df_silver.groupBy("pickup_day_of_week", "pickup_hour") \
                            .agg(
                                count("*").alias("total_de_corridas"),
                                round(sum("total_amount"), 2).alias("faturamento_total"),
                                round(avg("total_amount"), 2).alias("faturamento_medio_por_corrida"),
                                round(avg("trip_distance"), 2).alias("distancia_media_percorrida"),
                                round(avg("trip_duration_minutes"), 2).alias("duracao_media_da_viagem_minutos"),
                                round(avg("passenger_count"), 2).alias("media_de_passageiros"),
                                round(avg(col("trip_distance") / (col("trip_duration_minutes") / 60)), 2).alias("velocidade_media_kmh")
                            ).filter(col("duracao_media_da_viagem_minutos") > 0) # Adicionado filtro para evitar divisão por zero

df_gold_completa.write.mode("overwrite").option("overwriteSchema", "true").saveAsTable("workspace.taxi_nyc_schema.taxi_nyc_gold")

print("Tabela Gold 'taxi_nyc_gold' criada e pronta para análise.")

Tabela Gold 'taxi_nyc_gold' criada e pronta para análise.


- Métricas Adicionadas: Total de Corridas, Faturamento Total, Faturamento Médio por Corrida, Distância Média Percorrida, Duração Média da Viagem em Minutos, Média de Passageiros e Velocidade Média! 


- Total de Corridas por Hora (Demanda):


In [0]:

%sql
SELECT
  pickup_hour as Hora_Embarque,
  SUM(total_de_corridas) AS volume_total_corridas
FROM
  workspace.taxi_nyc_schema.taxi_nyc_gold
GROUP BY
  pickup_hour
ORDER BY
  pickup_hour


Hora_Embarque volume_total_corridas 0 2204 1 1477 2 998 3 658 4 476 5 647 6 1590 7 2919 8 3988 9 4509 10 4709 11 5045 12 5427 13 5760 14 5881 15 6053 16 6106 17 6455 18 6483 19 5534 20 5014 21 4852 22 4359 23 3122

In [0]:
%sql
SELECT
  pickup_hour,
  SUM(total_de_corridas) AS volume_total_corridas
FROM
  workspace.taxi_nyc_schema.taxi_nyc_gold
GROUP BY
  pickup_hour
ORDER BY
  pickup_hour

pickup_hour volume_total_corridas 0 2204 1 1477 2 998 3 658 4 476 5 647 6 1590 7 2919 8 3988 9 4509 10 4709 11 5045 12 5427 13 5760 14 5881 15 6053 16 6106 17 6455 18 6483 19 5534 20 5014 21 4852 22 4359 23 3122

Databricks visualization. Run in Databricks to view.

 - Faturamento por Dia da Semana

In [0]:
%sql
SELECT
  CASE
    WHEN pickup_day_of_week = 1 THEN '1 - Domingo'
    WHEN pickup_day_of_week = 2 THEN '2 - Segunda'
    WHEN pickup_day_of_week = 3 THEN '3 - Terça'
    WHEN pickup_day_of_week = 4 THEN '4 - Quarta'
    WHEN pickup_day_of_week = 5 THEN '5 - Quinta'
    WHEN pickup_day_of_week = 6 THEN '6 - Sexta'
    WHEN pickup_day_of_week = 7 THEN '7 - Sábado'
  END as dia_da_semana,
  SUM(faturamento_total) as faturamento
FROM
  taxi_nyc_schema.taxi_nyc_gold -- <-- NOME DO SCHEMA ATUALIZADO AQUI
GROUP BY
  pickup_day_of_week
ORDER BY
  faturamento DESC;


dia_da_semana faturamento 5 - Quinta 435108.76999999996 4 - Quarta 403499.92999999993 6 - Sexta 394801.41000000003 3 - Terça 383832.52 2 - Segunda 347733.4 7 - Sábado 333671.79999999993 1 - Domingo 290625.92

In [0]:
%sql
SELECT
  CASE
    WHEN pickup_day_of_week = 1 THEN '1 - Domingo'
    WHEN pickup_day_of_week = 2 THEN '2 - Segunda'
    WHEN pickup_day_of_week = 3 THEN '3 - Terça'
    WHEN pickup_day_of_week = 4 THEN '4 - Quarta'
    WHEN pickup_day_of_week = 5 THEN '5 - Quinta'
    WHEN pickup_day_of_week = 6 THEN '6 - Sexta'
    WHEN pickup_day_of_week = 7 THEN '7 - Sábado'
  END as dia_da_semana,
  SUM(faturamento_total) as faturamento
FROM
  taxi_nyc_schema.taxi_nyc_gold -- <-- NOME DO SCHEMA ATUALIZADO AQUI
GROUP BY
  pickup_day_of_week
ORDER BY
  faturamento DESC;

dia_da_semana faturamento 5 - Quinta 435108.76999999996 4 - Quarta 403499.92999999993 6 - Sexta 394801.41000000003 3 - Terça 383832.52 2 - Segunda 347733.4 7 - Sábado 333671.79999999993 1 - Domingo 290625.92

Databricks visualization. Run in Databricks to view.

- Dias Úteis vs. Fim de Semana

In [0]:
%sql
SELECT
  CASE 
    WHEN pickup_day_of_week IN (2, 3, 4, 5, 6) THEN 'Dia Útil'
    ELSE 'Fim de Semana'
  END AS tipo_de_dia,
  pickup_hour as hora_embarque,
  SUM(total_de_corridas) as total_corridas
FROM
  taxi_nyc_schema.taxi_nyc_gold
GROUP BY
  tipo_de_dia,
  pickup_hour
ORDER BY
  tipo_de_dia DESC;

tipo_de_dia hora_embarque total_corridas Fim de Semana 18 1525 Fim de Semana 14 1471 Fim de Semana 15 1489 Fim de Semana 0 1097 Fim de Semana 2 633 Fim de Semana 3 419 Fim de Semana 9 826 Fim de Semana 8 539 Fim de Semana 6 189 Fim de Semana 7 314 Fim de Semana 22 1097 Fim de Semana 20 1134 Fim de Semana 16 1625 Fim de Semana 5 95 Fim de Semana 21 1098 Fim de Semana 12 1414 Fim de Semana 10 1069 Fim de Semana 1 881 Fim de Semana 19 1355 Fim de Semana 11 1287 Fim de Semana 17 1561 Fim de Semana 4 230 Fim de Semana 13 1545 Fim de Semana 23 903 Dia Útil 18 4958 Dia Útil 7 2605 Dia Útil 8 3449 Dia Útil 16 4481 Dia Útil 14 4410 Dia Útil 2 365 Dia Útil 9 3683 Dia Útil 13 4215 Dia Útil 4 246 Dia Útil 6 1401 Dia Útil 15 4564 Dia Útil 1 596 Dia Útil 23 2219 Dia Útil 5 552 Dia Útil 17 4894 Dia Útil 21 3754 Dia Útil 19 4179 Dia Útil 20 3880 Dia Útil 22 3262 Dia Útil 3 239 Dia Útil 0 1107 Dia Útil 12 4013 Dia Útil 10 3640 Dia Útil 11 3758

In [0]:
%sql
SELECT
  CASE 
    WHEN pickup_day_of_week IN (2, 3, 4, 5, 6) THEN 'Dia Útil'
    ELSE 'Fim de Semana'
  END AS tipo_de_dia,
  pickup_hour as hora_embarque,
  SUM(total_de_corridas) as total_corridas
FROM
  taxi_nyc_schema.taxi_nyc_gold
GROUP BY
  tipo_de_dia,
  pickup_hour
ORDER BY
  tipo_de_dia DESC;

tipo_de_dia hora_embarque total_corridas Fim de Semana 18 1525 Fim de Semana 14 1471 Fim de Semana 15 1489 Fim de Semana 0 1097 Fim de Semana 2 633 Fim de Semana 3 419 Fim de Semana 9 826 Fim de Semana 8 539 Fim de Semana 6 189 Fim de Semana 7 314 Fim de Semana 22 1097 Fim de Semana 20 1134 Fim de Semana 16 1625 Fim de Semana 5 95 Fim de Semana 21 1098 Fim de Semana 12 1414 Fim de Semana 10 1069 Fim de Semana 1 881 Fim de Semana 19 1355 Fim de Semana 11 1287 Fim de Semana 17 1561 Fim de Semana 4 230 Fim de Semana 13 1545 Fim de Semana 23 903 Dia Útil 18 4958 Dia Útil 7 2605 Dia Útil 8 3449 Dia Útil 16 4481 Dia Útil 14 4410 Dia Útil 2 365 Dia Útil 9 3683 Dia Útil 13 4215 Dia Útil 4 246 Dia Útil 6 1401 Dia Útil 15 4564 Dia Útil 1 596 Dia Útil 23 2219 Dia Útil 5 552 Dia Útil 17 4894 Dia Útil 21 3754 Dia Útil 19 4179 Dia Útil 20 3880 Dia Útil 22 3262 Dia Útil 3 239 Dia Útil 0 1107 Dia Útil 12 4013 Dia Útil 10 3640 Dia Útil 11 3758

Databricks visualization. Run in Databricks to view.

- Piores Horários para Dirigir

In [0]:
%sql
SELECT
  pickup_hour as hora_embarque,
  AVG(duracao_media_da_viagem_minutos) AS duracao_media_min,
  AVG(velocidade_media_kmh) AS velocidade_media
FROM
  taxi_nyc_schema.taxi_nyc_gold
GROUP BY
  pickup_hour 
ORDER BY
  velocidade_media ASC;


hora_embarque duracao_media_min velocidade_media 17 17.965714285714288 8.938571428571429 15 20.29714285714286 8.995714285714284 13 18.97714285714286 9.291428571428572 12 18.99285714285714 9.367142857142856 18 15.800000000000002 9.447142857142856 16 20.25857142857143 9.448571428571428 14 19.860000000000003 9.530000000000001 10 17.83857142857143 10.298571428571426 19 14.981428571428571 10.537142857142857 9 17.265714285714285 10.932857142857143 8 17.53857142857143 11.182857142857143 20 14.614285714285712 11.314285714285715 21 14.855714285714285 11.865714285714287 22 14.855714285714285 12.417142857142858 23 15.124285714285714 13.994285714285715 6 19.33285714285714 15.127142857142857 0 14.01857142857143 15.247142857142858 1 13.598571428571429 15.739999999999998 2 13.062857142857142 15.811428571428573 5 21.192857142857143 17.677142857142858 3 13.644285714285715 18.235714285714288 11 18.457142857142856 30.16 7 18.09 53.01857142857143 4 19.562857142857144 154.3957142857143

In [0]:
%sql
SELECT
  pickup_hour as hora_embarque,
  AVG(duracao_media_da_viagem_minutos) AS duracao_media_min,
  AVG(velocidade_media_kmh) AS velocidade_media
FROM
  taxi_nyc_schema.taxi_nyc_gold
GROUP BY
  pickup_hour 
ORDER BY
  velocidade_media ASC;


hora_embarque duracao_media_min velocidade_media 17 17.965714285714288 8.938571428571429 15 20.29714285714286 8.995714285714284 13 18.97714285714286 9.291428571428572 12 18.99285714285714 9.367142857142856 18 15.800000000000002 9.447142857142856 16 20.25857142857143 9.448571428571428 14 19.860000000000003 9.530000000000001 10 17.83857142857143 10.298571428571426 19 14.981428571428571 10.537142857142857 9 17.265714285714285 10.932857142857143 8 17.53857142857143 11.182857142857143 20 14.614285714285712 11.314285714285715 21 14.855714285714285 11.865714285714287 22 14.855714285714285 12.417142857142858 23 15.124285714285714 13.994285714285715 6 19.33285714285714 15.127142857142857 0 14.01857142857143 15.247142857142858 1 13.598571428571429 15.739999999999998 2 13.062857142857142 15.811428571428573 5 21.192857142857143 17.677142857142858 3 13.644285714285715 18.235714285714288 11 18.457142857142856 30.16 7 18.09 53.01857142857143 4 19.562857142857144 154.3957142857143

Databricks visualization. Run in Databricks to view.

- Heatmap da Operação

In [0]:
%sql
SELECT
  pickup_hour,
  CASE
    WHEN pickup_day_of_week = 1 THEN '1-Dom'
    WHEN pickup_day_of_week = 2 THEN '2-Seg'
    WHEN pickup_day_of_week = 3 THEN '3-Ter'
    WHEN pickup_day_of_week = 4 THEN '4-Qua'
    WHEN pickup_day_of_week = 5 THEN '5-Qui'
    WHEN pickup_day_of_week = 6 THEN '6-Sex'
    WHEN pickup_day_of_week = 7 THEN '7-Sab'
  END as dia_da_semana,
  total_de_corridas
FROM
  taxi_nyc_schema.taxi_nyc_gold

Databricks visualization. Run in Databricks to view.

pickup_hour dia_da_semana total_de_corridas 4 5-Qui 66 2 7-Sab 322 23 6-Sex 637 7 7-Sab 155 8 6-Sex 601 9 2-Seg 665 7 1-Dom 159 3 2-Seg 66 22 7-Sab 690 3 7-Sab 220 15 7-Sab 812 6 2-Seg 243 18 3-Ter 952 9 4-Qua 814 7 4-Qua 562 3 6-Sex 68 3 3-Ter 29 2 4-Qua 49 20 6-Sex 731 4 4-Qua 36 3 4-Qua 39 8 5-Qui 759 15 1-Dom 677 0 5-Qui 251 9 3-Ter 765 5 3-Ter 101 17 4-Qua 1023 18 7-Sab 895 6 4-Qua 267 6 1-Dom 108 6 7-Sab 81 4 7-Sab 118 8 2-Seg 619 22 3-Ter 646 13 4-Qua 840 20 2-Seg 605 15 5-Qui 950 23 1-Dom 231 23 4-Qua 403 7 3-Ter 549 16 3-Ter 884 14 5-Qui 932 23 5-Qui 572 8 4-Qua 743 1 3-Ter 63 5 5-Qui 99 13 3-Ter 875 4 1-Dom 112 15 4-Qua 962 12 3-Ter 787 16 1-Dom 736 4 3-Ter 24 5 1-Dom 50 20 7-Sab 638 5 6-Sex 125 23 3-Ter 344 21 1-Dom 458 14 7-Sab 760 3 5-Qui 37 20 4-Qua 858 12 6-Sex 825 21 3-Ter 761 13 2-Seg 772 18 4-Qua 1022 19 6-Sex 851 13 1-Dom 731 19 3-Ter 812 0 7-Sab 544 1 7-Sab 419 4 2-Seg 56 1 6-Sex 186 20 3-Ter 780 10 3-Ter 720 21 6-Sex 697 13 7-Sab 814 0 3-Ter 172 7 5-Qui 540 23 2-Seg 263 22 1-Dom 407 21 7-Sab 640 0 1-Dom 553 19 4-Qua 925 9 5-Qui 742 14 2-Seg 837 0 2-Seg 206 19 5-Qui 943 18 6-Sex 1058 19 7-Sab 826 11 2-Seg 726 11 1-Dom 598 11 4-Qua 748 21 4-Qua 828 8 7-Sab 295 16 6-Sex 900 19 2-Seg 648 5 4-Qua 103 17 3-Ter 1059 4 6-Sex 64 17 5-Qui 1049 1 1-Dom 462 13 5-Qui 897 7 6-Sex 480 13 6-Sex 831 22 2-Seg 388 22 6-Sex 793 18 5-Qui 1133 11 7-Sab 689 1 2-Seg 129 17 1-Dom 681 11 6-Sex 727 5 7-Sab 45 8 3-Ter 727 14 1-Dom 711 21 5-Qui 880 2 2-Seg 105 10 1-Dom 488 9 1-Dom 404 20 1-Dom 496 20 5-Qui 906 1 5-Qui 134 8 1-Dom 244 0 4-Qua 159 7 2-Seg 474 10 7-Sab 581 9 6-Sex 697 2 1-Dom 311 2 5-Qui 66 14 3-Ter 860 17 7-Sab 880 2 6-Sex 114 11 5-Qui 790 21 2-Seg 588 16 4-Qua 901 17 2-Seg 785 12 1-Dom 662 15 2-Seg 826 18 2-Seg 793 2 3-Ter 31 12 2-Seg 751 11 3-Ter 767 12 4-Qua 826 9 7-Sab 422 22 5-Qui 808 10 4-Qua 764 23 7-Sab 672 5 2-Seg 124 18 1-Dom 630 10 5-Qui 783 3 1-Dom 199 17 6-Sex 978 14 6-Sex 884 15 6-Sex 917 6 3-Ter 322 0 6-Sex 319 10 2-Seg 675 12 5-Qui 824 22 4-Qua 627 16 5-Qui 984 19 1-Dom 529 10 6-Sex 698 6 5-Qui 310 6 6-Sex 259 14 4-Qua 897 1 4-Qua 84 16 7-Sab 889 15 3-Ter 909 12 7-Sab 752 16 2-Seg 812

## MACHINE LEARNING (AMOSTA 100MIL CORRIDAS):

#### Ensinar um modelo a prever a coluna trip_duration_minutes com base em outras características da corrida com SCIKIT-LEARN

**MODELO DE ML:** Regressão Linear


- Carregando os Dados e Convertendo para Pandas.

In [0]:
spark_df = spark.table("taxi_nyc_schema.taxi_nyc_silver").select(
    "pickup_hour",
    "pickup_day_of_week",
    "trip_distance",
    "passenger_count",
    "trip_duration_minutes"
).na.drop()

pandas_df = spark_df.limit(100000).toPandas()

- Preparando os Features:

In [0]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score
import numpy as np

X = pandas_df[["pickup_hour", "pickup_day_of_week", "trip_distance", "passenger_count"]]
y = pandas_df["trip_duration_minutes"]

- Dividindo o Teste e Treinando o Modelo de Regressão Linear:

In [0]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

lr_sklearn = LinearRegression()
lr_sklearn.fit(X_train, y_train)


LinearRegression()

- Previsão e Avaliação de Acurária:]

In [0]:
predictions = lr_sklearn.predict(X_test)

rmse = np.sqrt(mean_squared_error(y_test, predictions))
r2 = r2_score(y_test, predictions)


- Resultado da Regressão Linear (AMOSTRA):


In [0]:
print(f"Modelo Scikit-learn treinado!")
print(f"RMSE: {rmse:.2f} minutos")
print(f"R²: {r2:.2%}")

Modelo Scikit-learn treinado!
RMSE: 15.71 minutos
R²: -0.38%


RMSE: Ele pode errar até 15.71min pra mais ou pra menos do que a duração da viagem!

Um R² negativo (o seu caso) significa que o seu modelo é PIOR do que um simples chute na média. Ele está ativamente fazendo previsões ruins. POR ISSO, vamos treinar OUTRO modelo e avaliar novamente:


**MODELO DE ML:** Random Forest

n_estimators=50 significa que ele usará 50 "árvores de decisão" para votar na melhor previsão.
n_jobs=-1 usa todos os processadores disponíveis para acelerar o treinamento.

In [0]:
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_squared_error, r2_score
import numpy as np
rf_model = RandomForestRegressor(n_estimators=50, random_state=42, n_jobs=-1)

- Treindo Modelo:

In [0]:
rf_model.fit(X_train, y_train)

RandomForestRegressor(n_estimators=50, n_jobs=-1, random_state=42)

- Previsão e Verificando Acurácia:

In [0]:
predictions_rf = rf_model.predict(X_test)

rmse_rf = np.sqrt(mean_squared_error(y_test, predictions_rf))
r2_rf = r2_score(y_test, predictions_rf)

- Resultado da Random Forest (AMOSTRA):

In [0]:

print("\n--- Resultado do Random Forest ---")
print(f"RMSE: {rmse_rf:.2f} minutos")
print(f"R²: {r2_rf:.2%}")


--- Resultado do Random Forest ---
RMSE: 9.09 minutos
R²: 66.39%


%md
RMSE: Ele pode errar até 9.09min pra mais ou pra menos do que a duração da viagem!

Um R² de 66.39% significa que ele pode acertar 2/3 das previsões! Bem melhor do que o modelo de regressão linear para esse caso!